uvicorn/protocols/http/httptools_impl.py

```diff
@@ -245,7 +246,15 @@ def on_headers_complete(self) -> None:
         if http_version != "1.1":
             self.scope["http_version"] = http_version
         if self.parser.should_upgrade():
-            return
+            upgrade_http2 = False
+            for name, value in self.headers:
+                if name == b"upgrade":
+                    upgrade_value = value.lower()
+                    if upgrade_value == b"h2c":
+                        upgrade_http2 = True
+                        break
+            if not upgrade_http2:
+                return
```

Comment:
We can add a `self.is_http2`, and set as `True` on the `on_header` method.

Reply:
Oh indeed, the code looks like, we should also replace the `self.parser.should_upgrade()` in `on_body` and `on_message_complete` with `(self.parser.should_upgrade() and not self.is_http2)`. 
Sorry, I had overlooked that because no unit test failed for this edge case.

Reply:
We can probably benefit from more tests that consider those conditionals 🙌